R/data_processing_functions.R

```diff
@@ -1,34 +1,62 @@
 #' Compute growth of trees
 #'
-#' Based on two tree censuses compute the average annual growth in dbh for all
-#' trees that were alive at (TODO: fill this in).
+#' Based on two tree censuses, compute the average annual growth in `dbh` for all
+#' trees.
 #'
 #' @param census_1 A data frame of the first census.
 #' @param census_2 A data frame of the second (later) census
-#' @param id Name of variable that uniquely identifies each tree common to \code{census_1} and \code{census_2}
-#' allowing you to join/merge both data frames.
+#' @param id Name of variable that uniquely identifies each tree common
+#' to \code{census_1} and \code{census_2} allowing you to join/merge
+#' both data frames.
+#'
+#' @return An `sf` data frame with column \code{growth} giving the average
+#' annual growth in `dbh`.
 #'
-#' @return growth_df An sf data frame with \code{growth}: average annual growth in dbh.
-#' @export
 #' @import dplyr
 #' @import sf
+#'
+#' @family data processing functions
+#'
+#' @export
+#'
 #' @examples
 #' library(dplyr)
 #' library(stringr)
-#' # Load in data from two forst censuses
+#'
+#' # Load in data from two forest censuses
 #' data(census_1_ex, census_2_ex)
+#'
 #' # Filter out resprouts in second census
 #' census_2_ex_no_r <- census_2_ex %>%
 #'   filter(!str_detect(codes, "R"))
-#' id <- "ID"
-#' growth_ex <- compute_growth(census_1_ex, census_2_ex_no_r, id)
+#'
+#' growth_ex <- compute_growth(census_1_ex, census_2_ex_no_r, id = "ID")
 compute_growth <- function(census_1, census_2, id) {
+  # 0. Check inputs
+  check_inherits(census_1, "data.frame")
+  check_inherits(census_2, "data.frame")
+  check_inherits(id, "character")
+
+  if (
+    (!id %in% colnames(census_1)) | (n_distinct(census_1$ID) != nrow(census_1)) |
+    (!id %in% colnames(census_2)) | (n_distinct(census_2$ID) != nrow(census_2))
+  ){
+    glue_stop("The `id` argument must be the name of a numeric variable in both `census_1` and `census_2` that uniquely identifies each row.")
+  }
+
+  purrr::map2(
```

Comment:
Using the same reprex as above, the checking of `sp` being a factor is not working here. Is it because character is a form of factor?

``` r
library(tidyverse)
library(forestecology)
census_1_ex
#> # A tibble: 10 x 7
#>       ID sp                gx    gy date       codes   dbh
#>    <int> <chr>          <dbl> <dbl> <date>     <chr> <dbl>
#>  1     1 sugar maple     0.75  2.5  2015-06-01 M         5
#>  2     2 American beech  1.5   2.5  2015-06-01 M        20
#>  3     3 sugar maple     1.75  2.25 2015-06-01 M        15
#>  4     4 American beech  3     1.5  2015-06-01 M        12
#>  5     5 sugar maple     3.25  1.75 2015-06-01 M        35
#>  6     6 American beech  5.5   4.5  2015-06-01 M         6
#>  7     7 sugar maple     8     1.5  2015-06-01 M        22
#>  8     8 American beech  8.5   0.75 2015-06-01 M        14
#>  9     9 sugar maple     8.75  1.5  2015-06-01 M        42
#> 10    10 American beech  8.75  1.75 2015-06-01 M         4
```

Reply:
Currently, the inheritance checks are relaxed so that, if checking for factor, both character and factor are fine.

https://github.com/rudeboybert/forestecology/blob/99022eedfc44bd8770f28fe05734b09a63d9c99d/R/utils.R#L20-L28

Just pushed a commit that will be stricter about factors. :-)

Reply:
Ah I see, you set the terms in `utils.R`. very slick